framework/include/utils/ChainedADReal.h

```diff
@@ -0,0 +1,44 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ChainedReal.h"
+#include "ADRealForward.h"
+
+typedef DualNumber<ADReal, ADReal> ChainedADReal;
+
+/**
+ * We need to instantiate the following CompareTypes to tell the compiler that ADReal is a subtype
+ * of ChainedADReal. Otherwise specialization of operators on DualNumber<T, D> and DualNumber<T2,
+ * D2> is ambiguous.
+ */
+namespace MetaPhysicL
+{
+#define CompareTypes_default_Types(typenames, typename1, typename2, enabletype)                    \
+  CompareTypes_default_Type(Plus, typenames, typename1, typename2, enabletype);                    \
+  CompareTypes_default_Type(Minus, typenames, typename1, typename2, enabletype);                   \
+  CompareTypes_default_Type(Multiplies, typenames, typename1, typename2, enabletype);              \
+  CompareTypes_default_Type(Divides, typenames, typename1, typename2, enabletype);                 \
+  CompareTypes_default_Type(And, typenames, typename1, typename2, enabletype);                     \
+  CompareTypes_default_Type(Or, typenames, typename1, typename2, enabletype)
+
+#define CompareTypes_super(a, b, super)                                                            \
+  template <bool reverse_order, typename Enable>                                                   \
+  struct CompareTypes<a, b, reverse_order, Enable>                                                 \
+  {                                                                                                \
+    typedef super supertype;                                                                       \
+  };                                                                                               \
+  CompareTypes_default_Types(, a, b, void)
+
+CompareTypes_super(ADReal, ChainedADReal, ChainedADReal);
```

Comment:
It is by almost pure chance that this is working. The only reason you are not getting ambiguous `CompareTypes` instantiations is because implicitly you have `asd = false` in your `ChainedADReal` whereas `asd = true` for `ADReal`. That is the reason you are not ambiguous with respect to
```c++
template<typename T, typename D, typename T2, typename D2, bool asd>
struct CompareTypes<DualNumber<T, D, asd>, DualNumber<T2, D2, asd> > {
  typedef DualNumber<typename SymmetricCompareTypes<T, T2>::supertype,
                     typename SymmetricCompareTypes<
                       typename SymmetricCompareTypes<T, T2>::supertype,
                       typename SymmetricCompareTypes<D, D2>::supertype
                     >::supertype,
                    asd> supertype;
};
```
in `dualnumber_decl.h`. @roystgnr I would say this goes against what we've talked about for https://github.com/libMesh/MetaPhysicL/pull/36 wouldn't you?

Reply:
Moreover, it's not a good sign that this code only ever happened to compile because of chance ... unless @hugary1995 brilliantly realized that he wouldn't have to vie with any of the MetaPhysicL library overloads by having the template booleans be different...

Reply:
It worked by chance I'd say. Now I'll have to refresh my memory what asd stands for.

Reply:
Yeah, this is definitely unsupported.  `DualNumber` nested in `DualNumber` is in the test suite, which is why I overconfidently pointed you guys to it last year, but it works there (and doesn't need any further `CompareTypes` specialization there) because the derivative type is the same at every level.  We don't do non-symmetric Hessians, much less non-square.  I'm glad this hasn't broken horribly but if we want to spread the functionality everywhere we really ought to be figuring out how to do it with multiindices or something less fragile.